OpenRA.Game/Server/Server.cs

```diff
@@ -627,11 +676,119 @@ void DispatchOrdersToClient(Connection c, int client, int frame, byte[] data)
 			}
 		}
 
+		// Check if only one side remains (multiple players of a single team
+		// count as one, as long as it's not team 0)
+		bool HasRemainingConflict()
+		{
+			var teams = new HashSet<int>();
+			var remainingPlayers = gameInfo.Players.Where(p => p.Outcome == WinState.Undefined);
+
+			foreach (var player in remainingPlayers)
+			{
+				if (player.Team == 0 && teams.Contains(0))
+					return true;
+
+				teams.Add(player.Team);
+				if (teams.Count > 1)
+					return true;
+			}
+
+			return false;
+		}
+
+		void SetRemainingPlayersAsWinners()
+		{
+			var now = DateTime.UtcNow;
+			var remainingPlayers = gameInfo.Players.Where(p => p.Outcome == WinState.Undefined);
+
+			foreach (var winner in remainingPlayers)
+			{
+				winner.Outcome = WinState.Won;
+				winner.OutcomeTimestampUtc = now;
+			}
+		}
+
+		void SetPlayerDefeat(int client)
+		{
+			var defeatedPlayer = gameInfo.Players.FirstOrDefault(p => p.ClientIndex == client);
+			if (defeatedPlayer == null || defeatedPlayer.Outcome != WinState.Undefined)
+				return;
+
+			defeatedPlayer.Outcome = WinState.Lost;
+			defeatedPlayer.OutcomeTimestampUtc = DateTime.UtcNow;
+
+			if (!HasRemainingConflict())
+				SetRemainingPlayersAsWinners();
+		}
+
+		bool OutOfSync(int frame)
+		{
+			Log.Write("server", "Out of sync detected at frame {0}, cancel replay recording", frame);
+			recorder.Metadata = null; // make sure the written file is not valid
```

Comment:
Style nit (here and below): Comments should go on their own line preceding the line commented on and start with a capital letter.

Reply:
Done